api/src/main/java/org/openmrs/api/impl/ObsServiceImpl.java

```diff
@@ -87,7 +87,148 @@ public void onShutdown() {
 	 * @see org.openmrs.api.ObsService#saveObs(org.openmrs.Obs, String)
 	 */
 	public Obs saveObs(Obs obs, String changeMessage) throws APIException {
-		if (null != obs && null != obs.getConcept() && obs.getConcept().isComplex()
+		if(obs == null){
+			throw new APIException("obs cannot be null!!", (Object[]) null); //TODO: change this to a message string!!
+		}
+
+		if(obs.getId() != null && changeMessage == null){
+			throw new APIException("Obs.error.ChangeMessage.required", (Object[]) null);
+		}
+
+		handleExistingObsWithComplexConcept(obs);
+
+		ensureRequirePrivilege(obs);
+
+		//Should allow updating a voided Obs, it seems to be pointless to restrict it,
+		//otherwise operations like merge patients won't be possible when to moving voided obs
+		if (obs.getObsId() == null || obs.getVoided()) {
+			return saveNewAndVoidedObs(obs,changeMessage);
+		} else if(!obs.isDirty()){
+			setPersonFromEncounter(obs);
```

Comment:
I find it strange that if the Obs is not dirty, that you're setting the person, should this logic really be in Encounter.java when you add an Obs? And may be in Obs.java when a group member is added